Fix bug and update dockerfile

ABaldwinHunter opened the pull request:
@codeclimate/review 

Fixes a bug where `gofmt` didn't get called correctly, resulting in no issues being produced.

See individual commits for more detail.

In a future PR, I'd like to add a test suite.

Commit:
Update dockerfile

- Use newer golang image - old one no longer exists
- Make dockerfile compliant with spec

Commit:
Add support for logging errors

Commit:
Correctly call gofmt

Fixes bug where binary was no longer available in project

Commit:
Correctly pluralize 'places' in issue description

gdiggs commented on Dockerfile:
```diff
@@ -3,9 +3,9 @@ FROM golang:alpine
 
 RUN adduser -u 9000 -D app
-ADD build/codeclimate-gofmt /usr/src/app
+COPY . /usr/src/app
```
What does this change do to the size of the image? Did you try doing `COPY build/codeclimate-gofmt /usr/src/app` instead of `ADD`?

ABaldwinHunter replied:
@GordonDiggs thanks good idea! Updated to only copy the binary.

The image size is still bigger than before - probably due to using the larger golang image instead of alpine*:

```
codeclimate/codeclimate-gofmt      latest     80a2cca69578        About a minute ago   247.6 MB
codeclimate/codeclimate-gofmt      newer     376f5bbe035a        8 hours ago          270.5 MB
codeclimate/codeclimate-gofmt       old       8b5b9e71bd0e        6 months ago         218.3 MB
```

Currently looking at the circle failure. Engine works locally. 

I'm suspicious of 

```
stat /var/lib/docker/btrfs/subvolumes/7c5f9c8c0809d25df2ac0fbad945c4d8724b87ae197dd269da8dfaaa0b9b8998/
```

 before /usr/src/app/codeclimate-gofmt (edit: in the circle build failure)

\* `yunspace/alpine-golang` no longer exists. There's also a `golang:alpine` image I considered, but ran into an issue missing git during build script, and also their docs suggest golang:latest might be the way to go for long lived projects: https://hub.docker.com/_/golang/

wfleming commented on Dockerfile:
```diff
@@ -1,8 +1,11 @@
+FROM golang:alpine
 
-ADD build/codeclimate-gofmt /usr/src/app/
+WORKDIR /usr/src/app
```
Having the work dir in the engine's source dir is unusual for us. Should we move the work dir to `/code` later on in this file?

ABaldwinHunter replied:
I like that

ABaldwinHunter replied:
Added below.

pbrisbin replied:
FYI: having `/code` be the WORKDIR is a requirement of the SPEC. Maybe you should double check that this Dockerfile is in accordance in general?

ABaldwinHunter replied:
@pbrisbin Good point. Checked out the spec and updated to include a maintainer.

ABaldwinHunter commented on bin/build:
```diff
@@ -13,6 +13,6 @@ BUILD_SCRIPT="cd /src && "\
 "go build -o build/codeclimate-gofmt ."
 
-docker run -v "$SRC_ROOT:/src" yunspace/alpine-golang /bin/sh -c "$BUILD_SCRIPT"
+docker run -v "$SRC_ROOT:/src" golang:latest /bin/sh -c "$BUILD_SCRIPT"
```
note that this script is using the full golang image, but the dockerfile has the `golang:alpine` image. (The full image seems to have git support for grabbing packages)

gdiggs replied:
Make sure this compiled binary works on the resulting image

ABaldwinHunter replied:
Thanks. Yep engine runs!

ABaldwinHunter commented:
@GordonDiggs @wfleming Updated! Ready for another look here. Will squash docker commits into first if all looks okay.

pbrisbin commented on Dockerfile:
```diff
@@ -1,8 +1,17 @@
+WORKDIR /code
+
+COPY build/ /usr/src/app
+COPY build/codeclimate-gofmt /usr/src/app/build/codeclimate-gofmt
```
If you've copied all of `build/` into `/usr/src/app`, that has already put `/usr/src/app/codeclimate-gofmt` in place. Why not just use that, instead of making another copy in a `build` subdirectory?

pbrisbin commented on Dockerfile:
```diff
@@ -1,8 +1,17 @@
+
+COPY build/ /usr/src/app
+COPY build/codeclimate-gofmt /usr/src/app/build/codeclimate-gofmt
+RUN chown -R app:app /usr/src/app
```
Discussed in slack, but I'm not sure this is needed. Does the engine work without it?

ABaldwinHunter replied:
It does. Got rid of that and also the `USER app` line which isn't in the spec example and wasn't needed for engine to work locally.

pbrisbin commented:
Do your changes make it so that, if the same or a similar bug is re-introduced, this engine behaves better than silently exiting 0 now?

ABaldwinHunter commented:
@pbrisbin @GordonDiggs @wfleming Thanks for feedback! 

Made Dockerfile spec compliant and rebased commits. Ready for 🔎

ABaldwinHunter commented on codeclimate-gofmt.go:
```diff
@@ -1,27 +1,33 @@
 
 	for _, path := range analysisFiles {
-		cmd := exec.Command("/usr/src/app/bin/gofmt", "-d", path)
+		cmd := exec.Command("gofmt", "-d", path)
```
The binary `"/usr/src/app/bin/gofmt"` isn't available in image

pbrisbin replied:
Just to confirm, with your change at lines 33-35, if you revert this the engine exits non-zero with that output? If so, then :+1:

ABaldwinHunter replied:
yup

```
$ codeclimate analyze -e gofmt
Starting analysis
Running gofmt: Done!
error: (CC::Analyzer::Engine::EngineFailure) engine gofmt:stable failed with status 1 and stderr 
Error analyzing path: /code/punch/help_command.go
Error: fork/exec bin/gofmt: no such file or directory
```

ABaldwinHunter commented on codeclimate-gofmt.go:
```diff
@@ -1,27 +1,33 @@
 
 	for _, path := range analysisFiles {
-		cmd := exec.Command("/usr/src/app/bin/gofmt", "-d", path)
+		cmd := exec.Command("gofmt", "-d", path)
```
This change is probably the root of the original bug

ABaldwinHunter commented:
@pbrisbin: it doesn't. Right now, there are a couple of places in the main script where an error result in exit(1), e.g.: `LoadConfig()`.

(edit: example https://github.com/codeclimate-community/codeclimate-gofmt/blob/master/codeclimate-gofmt.go#L20)

Having less context around the go engine behavior, it seemed very intentional that similar exits weren't introduced in other places that we checked for errors:

https://github.com/codeclimate-community/codeclimate-gofmt/pull/9/files#diff-e01f8d974bc50ae387673a755561b607R33

As a simple first step, it would be easy to add `exit(1)` in those spots as well.

Potential downside: noise from errors that seem to have been intentionally skipped in past

gdiggs commented:
I think we should exit at those places we are currently not

Commit:
Exit on error

Avoid failing silently - which allowed a bug to creep in.
Sometimes the gofmt command will error because of something in the code it's
analyzing, like a parsing error.

In this case, it's good to blow up because the analysis doesn't continue.
In this case we should add the gofmt output to stderr for more information.

pbrisbin commented on Dockerfile:
```diff
@@ -1,8 +1,14 @@
+WORKDIR /code
+
+COPY build/ /usr/src/app
+COPY build/codeclimate-gofmt /usr/src/app/build/codeclimate-gofmt
```
My comment got lost in out-of-date-diff land, so re-adding:

If you're already copying all of `build/` above, that means `/usr/src/app/codeclimate-gofmt` will exist. Why not use that instead of making a redundant layer to put it at a different location?

wfleming replied:
I feel like we shouldn't actually need any contents of `build` other than the bin? That was the intent of the older `ADD` ccommand. Could we just have this last COPY without the one before it?

pbrisbin replied:
:+1: with the minor nit that it doesn't need the `build/` parent directory, just put it at `/usr/src/app/codeclimate-gofmt`

ABaldwinHunter commented:
@GordonDiggs yeah, agree. Would have been a better side to err on. Updated.

Explained in commit, but a parsing error will also have that information output. Example:

```
$ codeclimate analyze -e gofmt timecard/
Starting analysis
Running gofmt: Done!
error: (CC::Analyzer::Engine::EngineFailure) engine gofmt:stable failed with status 1 and stderr 
Analyzing path: /code/timecard/entry.go
Error analyzing path: /code/timecard/entry.go
Error: exit status 2
Gofmt output: /code/timecard/entry.go:9:1: illegal character U+0023 '#'
```

pbrisbin commented on Dockerfile:
```diff
@@ -1,8 +1,13 @@
+MAINTAINER Code Climate <hello@codeclimate.com>
 
 RUN adduser -u 9000 -D app
-USER app
```
This is in the SPEC and should be preserved

> A non-root user named app must be created with UID and GID 9000 and declared using the USER directive.

https://github.com/codeclimate/spec/blob/master/SPEC.md#packaging

wfleming commented on Dockerfile:
```diff
@@ -1,8 +1,13 @@
+
+WORKDIR /code
+
+COPY build /usr/src/app
```
Comment chain got lost in a diff again.

Do we actually need anything in `build` other than the `codeclimate-gofmt` bin? If not, can we just copy that bin and not whatever other junk also ends up in `build`?

ABaldwinHunter replied:
@wfleming In some earlier commits, for some reason - it didn't work to have

```
COPY build/codeclimate-gofmt /usr/src/app
```

in the dockerfile, with

```
CMD ["/usr/src/app/codeclimate-gofmt"]
```

https://circleci.com/gh/codeclimate-community/codeclimate-gofmt/26

> Step 3 : COPY build/codeclimate-gofmt /usr/src/app
> stat /var/lib/docker/btrfs/subvolumes/6cd501cd74c8efd4100d078edaa3e1aa6d90339ac71c1054edf62deb7a2e32a6/usr/src/app/codeclimate-gofmt: not a directory

wfleming replied:
You may need to trailing slash?

```
COPY build/codeclimate-gofmt /usr/src/app
```

ABaldwinHunter replied:
oo yesssss. thanks!

ABaldwinHunter commented:
@GordonDiggs @pbrisbin @wfleming Updated dockerfile and added exits. Ready for a final look!

pbrisbin commented:
LGTM

On the testing front, I'd recommend a basic integration using cram that runs the engine on an example file and asserts it outputs the expected issues. I'd be in favor of that as a general engines' test suite pattern.

ABaldwinHunter commented:
@pbrisbin Cram looks cool! https://bitheap.org/cram/
